### ISSUE TITLE
End of document is cut off when multi-byte UTF-8 characters are used

### ISSUE THREAD

**Issue by roastchicken** [wereturtle/ghostwriter#699]
## Platform & version

**Platform:** Windows 10 Home 21H1, build 19043.1348

**ghostwriter installation:**
 - ghostwriter version: 2.1.0 (also occurred in 2.0.2)
 - installation type: [ghostwriter_x64_portable.7z from michelovera](https://github.com/michelolvera/vs-ghostwriter/releases/tag/win-2.1.0)

## Description
Every time a non-ASCII character<sup>[1]</sup> is entered, the end of the document in the preview is truncated by one or more bytes.

## Steps to Reproduce
1. Start ghostwriter
2. Create a new file (*File -> New*, *Ctrl + N*)
3. Turn on live preview (*View -> Preview in HTML*, *Ctrl + P*)
4. Type any ASCII characters
5. Type a multi-byte character anywhere **before** the ASCII characters (if you need one, copy this: `å` <sup>[2]</sup>)
6. Observe that the last character(s) disappear from the preview (depending on the number of bytes in the character entered in the previous step)

## Expected behavior
I expect ghostwriter to preview the entire file regardless of how many multi-byte characters are entered.

### Ghostwriter
![ghostwriter_2021-12-08_14-49-25](https://user-images.githubusercontent.com/524709/145219641-861d1652-29e1-40c1-8711-2bcd001bb5d6.png)

### Notepad++ with special characters displayed
![notepad++_2021-12-08_14-53-30](https://user-images.githubusercontent.com/524709/145220377-70b93414-8467-4a63-aab9-b6c56d47c65f.png)

## Test markdown
[multi-byte-truncation_test-case.md](https://github.com/wereturtle/ghostwriter/files/7676459/multi-byte-truncation_test-case.md)

## Additional context

I suspect that this may very well be an issue with cmark-gfm, but I have no way of testing this _immediately_ because I am unfamiliar with compiling on Windows and don't have any other OSes to test on. Skimming through the source code of ghostwriter, I could not figure out what version of cmark-gfm is used other than that it was added over two years ago. Perhaps cmark-gfm has fixed this bug in an update, in which case ghostwriter should update the bundled cmark-gfm.

## Footnotes
- [1]: My knowledge of Unicode is limited, but as far as I can ascertain all single-byte characters (in UTF-8) are also ASCII characters because Unicode is a superset of [US-ASCII](https://en.wikipedia.org/wiki/ASCII). The first sentence of [this Wikipedia article](https://en.wikipedia.org/w/index.php?title=Basic_Latin_(Unicode_block)&oldid=1044748335) seems to agree with me.
- [2]: I use Codepoints.net for Unicode character information, for example [here is its page about `å`](https://codepoints.net/U+00E5).

**Comment by veterinarius-de** [comment]
I can confirm this issue on Windows Edition: Windows 10 Pro, Version: 21H2, German Language.
The program works fine on ubuntu 21.10, German Language. All the characters are displayed at the preview.

**Comment by roastchicken** [comment]
> I can confirm this issue on Windows Edition: Windows 10 Pro, Version: 21H2, German Language. The program works fine on ubuntu 21.10, German Language. All the characters are displayed at the preview.

@veterinarius-de what type of Windows installation are you using (in other words: is it a portable or full install)? And where did you download it from? It would also be helpful if you could provide which version(s) of ghostwriter you tested on.

Because veterinarius-de was able to reproduce this issue on Windows but not Ubuntu, my suspicion of this being an issue with cmark-gfm grows. Perhaps michelolvera's build process has caused this bug?

I have some free time now, so I'll try to compile cmark/cmark-gfm on Windows to see if I can isolate the source of this bug.

**Comment by veterinarius-de** [comment]
@roastchicken - Version of Ghostwriter at ubuntu: 1.8.1 (https://packages.ubuntu.com/impish/ghostwriter) :-(
I use the portable Version of Ghostwriter at Windows: 2.1.0
After Install from https://wereturtle.github.io/ghostwriter/download.html#linux at ubuntu: 2.0.1 - it works as expected (without the issue) but the version numbers are still different

**Comment by wereturtle** [comment]
Yeah I can't replicate this on Fedora.  Hmm.  Out of curiosity, does Pandoc or MultiMarkdown work for you?

**Comment by Cxarli** [comment]
FYI, running the program with the Address Sanitizer turned on, does yield some out-of-bounds reads coming from cmark-gfm, even on Linux:

<details>

```
ERROR: AddressSanitizer: heap-buffer-overflow
    #0  in S_parser_feed 3rdparty/cmark-gfm/core/blocks.c:700
    #1  in cmark_parser_feed 3rdparty/cmark-gfm/core/blocks.c:669
    #2  in ghostwriter::CmarkGfmAPI::parse(QString const&, bool) src/cmarkgfmapi.cpp:96
    #3  in ghostwriter::MarkdownEditorPrivate::parseDocument() src/markdowneditor.cpp:1745
    #4  in ghostwriter::MarkdownEditor::onContentsChanged(int, int, int) src/markdowneditor.cpp:1514
```

</details>

Taking a closer look at our code, we do:

`cmark_parser_feed(parser, text.toLatin1().data(), text.toLocal8Bit().length());`

Note that we use the data in Latin-1 but the length in UTF-8. Testing with `text = QString("œ")` this gives:

```
Latin: 3f (length = 1)
Loc8b: c5 93 (length = 2)
```

Which means the length() is longer than the data, causing all types of weird issues. Now normally I would suggest setting both to `toLocal8Bit`, but this line of code has the following warning above it:

```
// Use Latin1 instead of UTF-8 since the column numbers and even
// the nodes returned in the AST are shifted or missing when
// UTF-8 characters longer than 1 byte are encountered.
```

So this may not be trivial to fix.

------------

To reproduce it on Linux, try filling your document with the 𒂟 character, which is 0x3F 0x3F (double question mark) in Latin-1 and 0xF0 0x92 0x82 0x9F in UTF-8. This means we will try to read twice as many bytes as there are in the buffer and hit a segfault:

```
Latin: 3f 3f 3f 3f (length = 13312)
Loc8b: f0 92 82 9f (length = 26624)
AddressSanitizer:DEADLYSIGNAL
=================================================================
ERROR: AddressSanitizer: SEGV on unknown address 
The signal is caused by a READ memory access.
    #0  in S_parser_feed 3rdparty/cmark-gfm/core/blocks.c:700
    #1  in cmark_parser_feed 3rdparty/cmark-gfm/core/blocks.c:669
    #2  in ghostwriter::CmarkGfmAPI::parse(QString const&, bool) src/cmarkgfmapi.cpp:96
```

The reason this doesn't show up visually in the render for you, is probably because Linux is a bit more lenient with out-of-bounds reads and writes. Windows may be stricter on this, but the problem is universal.

**Comment by roastchicken** [comment]
> Out of curiosity, does Pandoc or MultiMarkdown work for you?

Pandoc works as expected for my test case using [Pandoc Markdown](https://pandoc.org/MANUAL.html#pandocs-markdown), CommonMark Markdown, and GitHub-Flavored Markdown.

**Comment by wereturtle** [comment]
This should be fixed in version 2.1.2.  Please let me know if the issue persists.  Thanks!
